File: app/src/main/java/com/babestudios/companyinfouk/ui/search/SearchResultsAdapter.java
package com.babestudios.companyinfouk.ui.search;


import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.babestudios.companyinfouk.R;
import com.babestudios.companyinfouk.data.model.search.CompanySearchResult;
import com.babestudios.companyinfouk.data.model.search.CompanySearchResultItem;

import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import rx.Observable;
import rx.schedulers.Schedulers;

public class SearchResultsAdapter extends RecyclerView.Adapter<SearchResultsAdapter.SearchResultsViewHolder> {

	private SearchResultsRecyclerViewClickListener mItemListener;

	private CompanySearchResult companySearchResults = new CompanySearchResult();
	private List<CompanySearchResultItem> filteredSearchResults;

	SearchResultsAdapter(Context c, CompanySearchResult companySearchResults, SearchPresenter.FilterState filterState) {
		mItemListener = (SearchResultsRecyclerViewClickListener) c;
		this.companySearchResults = companySearchResults;
		updateFilteredResults(companySearchResults, filterState);
	}

	private void updateFilteredResults(CompanySearchResult companySearchResults, SearchPresenter.FilterState filterState) {
		if (filterState.ordinal() > SearchPresenter.FilterState.FILTER_SHOW_ALL.ordinal()) {
			Observable.from(companySearchResults.items).filter(companySearchResultItem -> companySearchResultItem.companyStatus.equalsIgnoreCase(filterState.toString()))
					.observeOn(Schedulers.immediate())
					.toList()
					.subscribe(result -> filteredSearchResults = result);
		} else {
			filteredSearchResults = companySearchResults.items;
		}
	}

	@Override
	public SearchResultsViewHolder onCreateViewHolder(ViewGroup parent, int i) {
		View itemLayoutView = LayoutInflater.from(parent.getContext())
				.inflate(R.layout.search_result_list_item, parent, false);

		return new SearchResultsViewHolder(itemLayoutView);
	}

	@Override
	public void onBindViewHolder(SearchResultsViewHolder viewHolder, int position) {
		viewHolder.lblCompanyName.setText(filteredSearchResults.get(position).title);
		viewHolder.lblAddress.setText(filteredSearchResults.get(position).addressSnippet);
		viewHolder.lblActiveStatus.setText(filteredSearchResults.get(position).companyStatus);
		viewHolder.lblIncorporated.setText(filteredSearchResults.get(position).description);
	}

	public long getItemId(int position) {
		return position;
	}

	@Override
	public int getItemCount() {
		return filteredSearchResults.size();
	}

	void setFilterOnAdapter(SearchPresenter.FilterState filterState) {
		updateFilteredResults(companySearchResults, filterState);
		notifyDataSetChanged();
	}

	class SearchResultsViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
		@Bind(R.id.lblAddress)
		TextView lblAddress;
		@Bind(R.id.lblCompanyName)
		TextView lblCompanyName;
		@Bind(R.id.lblActiveStatus)
		TextView lblActiveStatus;
		@Bind(R.id.lblIncorporated)
		TextView lblIncorporated;

		SearchResultsViewHolder(View itemView) {
			super(itemView);
			ButterKnife.bind(this, itemView);
			itemView.setOnClickListener(this);
		}

		@Override
		public void onClick(View v) {
			mItemListener.searchResultItemClicked(v, this.getLayoutPosition(), filteredSearchResults.get(getLayoutPosition()).title, filteredSearchResults.get
					(getLayoutPosition()).companyNumber);
		}
	}

	interface SearchResultsRecyclerViewClickListener {
		void searchResultItemClicked(View v, int position, String companyName, String companyNumber);
	}

	void addItems(CompanySearchResult companySearchResults, SearchPresenter.FilterState filterState) {
		this.companySearchResults.items.addAll(companySearchResults.items);
		updateFilteredResults(companySearchResults, filterState);
		notifyDataSetChanged();
	}
}

File: app/src/main/java/com/babestudios/companyinfouk/ui/search/SearchResultsAdapter.kt
package com.babestudios.companyinfouk.ui.search

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.babestudios.companyinfouk.R
import com.babestudios.companyinfouk.data.model.search.CompanySearchResult
import com.babestudios.companyinfouk.data.model.search.CompanySearchResultItem

import butterknife.BindView
import butterknife.ButterKnife
import com.babestudios.base.ext.biLet
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers

class SearchResultsAdapter internal constructor(c: Context, private val companySearchResults: CompanySearchResult, filterState: SearchPresenter.FilterState) : RecyclerView.Adapter<SearchResultsAdapter.SearchResultsViewHolder>() {

	private val mItemListener: SearchResultsRecyclerViewClickListener

	private var filteredSearchResults: List<CompanySearchResultItem>? = null

	init {
		mItemListener = c as SearchResultsRecyclerViewClickListener
		updateFilteredResults(companySearchResults, filterState)
	}

	private fun updateFilteredResults(companySearchResults: CompanySearchResult, filterState: SearchPresenter.FilterState) {
		if (filterState.ordinal > SearchPresenter.FilterState.FILTER_SHOW_ALL.ordinal) {
			Observable.fromIterable(companySearchResults.items)
					.filter { companySearchResultItem -> companySearchResultItem.companyStatus != null && companySearchResultItem.companyStatus.equals(filterState.toString(), ignoreCase = true) }
					.observeOn(Schedulers.trampoline())
					.toList()
					.subscribe { result -> filteredSearchResults = result }
		} else {
			filteredSearchResults = companySearchResults.items
		}
	}

	override fun onCreateViewHolder(parent: ViewGroup, i: Int): SearchResultsViewHolder {
		val itemLayoutView = LayoutInflater.from(parent.context)
				.inflate(R.layout.row_search_result, parent, false)

		return SearchResultsViewHolder(itemLayoutView)
	}

	override fun onBindViewHolder(viewHolder: SearchResultsViewHolder, position: Int) {
		filteredSearchResults?.let {
			viewHolder.lblCompanyName?.text = it[position].title
			viewHolder.lblAddress?.text = it[position].addressSnippet
			viewHolder.lblActiveStatus?.text = it[position].companyStatus
			viewHolder.lblIncorporated?.text = it[position].description
		}
	}

	override fun getItemId(position: Int): Long {
		return position.toLong()
	}

	override fun getItemCount(): Int {
		return filteredSearchResults?.size ?: 0
	}

	internal fun setFilterOnAdapter(filterState: SearchPresenter.FilterState) {
		updateFilteredResults(companySearchResults, filterState)
		notifyDataSetChanged()
	}

	inner class SearchResultsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
		@JvmField
		@BindView(R.id.lblAddress)
		var lblAddress: TextView? = null
		@JvmField
		@BindView(R.id.lblCompanyName)
		var lblCompanyName: TextView? = null
		@JvmField
		@BindView(R.id.lblActiveStatus)
		var lblActiveStatus: TextView? = null
		@JvmField
		@BindView(R.id.lblIncorporated)
		var lblIncorporated: TextView? = null

		init {
			ButterKnife.bind(this, itemView)
			itemView.setOnClickListener(this)
		}

		override fun onClick(v: View) {
			filteredSearchResults?.let {
				(it[layoutPosition].title to it[layoutPosition].companyNumber).biLet {title, companyNumber->
					mItemListener.searchResultItemClicked(v, this.layoutPosition, title, companyNumber)
				}
			}
		}
	}

	internal interface SearchResultsRecyclerViewClickListener {
		fun searchResultItemClicked(v: View, position: Int, companyName: String, companyNumber: String)
	}

	internal fun addItems(companySearchResults: CompanySearchResult, filterState: SearchPresenter.FilterState) {
		this.companySearchResults.items.addAll(companySearchResults.items)
		updateFilteredResults(companySearchResults, filterState)
		notifyDataSetChanged()
	}
}
